### ISSUE TITLE
CookieContainer doesn't return all relevant cookies

### ISSUE THREAD

**Issue by jpelcis** [dotnet/corefx#24368]
Adding a new non-matching cookie can prevent matching cookies from being found.

```c#
var uri1 = new Uri("http://www.example.com/");
var uri2 = new Uri("http://www.example.com/level1/");
var uri3 = new Uri("http://www.example.com/level1/level2a");
var uri4 = new Uri("http://www.example.com/level1/level2b");
var cookieContainer = new CookieContainer();

cookieContainer.Add(new Cookie("cookie1", "value1", uri1.AbsolutePath, uri1.Host));
cookieContainer.Add(new Cookie("cookie2", "value2", uri2.AbsolutePath, uri2.Host));
cookieContainer.Add(new Cookie("cookie3", "value3", uri3.AbsolutePath, uri3.Host));

var beforeCount = cookieContainer.GetCookies(uri3).Count;
Console.WriteLine(beforeCount); // 3

cookieContainer.Add(new Cookie("cookie4", "value4", uri4.AbsolutePath, uri4.Host));

var afterCount = cookieContainer.GetCookies(uri3).Count;
Console.WriteLine(afterCount); // 2
```

**Comment by karelz** [comment]
Test added in PR #24706.

@davidsh what is the next step?
We accepted the "failing" test in #24706. Are we ready to make progress on a fix in PR #24369? Or is more discussion needed?

**Comment by davidsh** [comment]
>@davidsh what is the next step?

PR #24369 will have outerloop tests run first.  Then I think it will be ready to merge.
